File: sdk/kotlin/src/main/kotlin/com/gosslens/AudioOutput.kt
package com.gosslens

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import android.os.Build
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Routes the lens mixer to the device speaker. The pull is graph-thread
 * only, so [pump] runs in the frame loop and fills a ring a writer
 * thread drains into a streaming [AudioTrack]; underrun plays silence.
 */
class GossAudioOutput(private val session: GossSession) {
    /** The state the writer thread and the drop safety net share. It holds
     * no reference to the wrapper: a running thread is a GC root, so a
     * thread capturing the wrapper would pin it (and its session) forever
     * and the platform track could never be reclaimed. */
    private class Playback {
        val ring = ShortArray(RING_FRAMES)
        var readIndex = 0
        var writeIndex = 0
        val lock = Object()
        val chunk = ShortArray(CHUNK_FRAMES)
        var track: AudioTrack? = null
        var writer: Thread? = null
        @Volatile var running = false

        fun stop() {
            running = false
            writer?.join(500)
            writer = null
            track?.stop()
            track?.release()
            track = null
        }
    }

    private val playback = Playback()
    private val pullBuffer: ByteBuffer =
        ByteBuffer.allocateDirect(PULL_FRAMES * 2).order(ByteOrder.nativeOrder())
    // One view over the pull buffer, created once so pump() allocates
    // nothing per frame.
    private val pullShorts = pullBuffer.asShortBuffer()

    /** Releases the track and thread if the wrapper is dropped without
     * stop(); the action holds only the playback state, never the wrapper. */
    private val cleanable: Any? =
        if (Build.VERSION.SDK_INT >= 33) NativeCleaner.register(this, PlaybackStopper(playback)) else null

    private class PlaybackStopper(private val playback: Playback) : Runnable {
        override fun run() = playback.stop()
    }

    /** Starts the platform track and its writer thread. */
    fun start() {
        if (playback.track != null) return
        val minBytes = AudioTrack.getMinBufferSize(
            SAMPLE_RATE, AudioFormat.CHANNEL_OUT_MONO, AudioFormat.ENCODING_PCM_16BIT,
        )
        val t = AudioTrack.Builder()
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_MEDIA)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .build(),
            )
            .setAudioFormat(
                AudioFormat.Builder()
                    .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                    .setSampleRate(SAMPLE_RATE)
                    .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                    .build(),
            )
            .setBufferSizeInBytes(maxOf(minBytes, CHUNK_FRAMES * 4))
            .setTransferMode(AudioTrack.MODE_STREAM)
            .build()
        t.play()
        val p = playback
        p.track = t
        p.running = true
        p.writer = Thread {
            while (p.running) {
                var filled = 0
                synchronized(p.lock) {
                    while (filled < p.chunk.size && p.readIndex != p.writeIndex) {
                        p.chunk[filled] = p.ring[p.readIndex]
                        p.readIndex = (p.readIndex + 1) % RING_FRAMES
                        filled += 1
                    }
                }
                while (filled < p.chunk.size) {
                    p.chunk[filled] = 0
                    filled += 1
                }
                t.write(p.chunk, 0, p.chunk.size)
            }
        }.also { it.name = "goss-audio-out"; it.start() }
    }

    /**
     * Pulls the next mixer block into the ring. Call once per frame
     * from the same thread that ticks the lens.
     */
    fun pump(frames: Int = 800) {
        val count = minOf(frames, PULL_FRAMES)
        pullBuffer.clear()
        if (!session.pullAudio(pullBuffer, count)) return
        val p = playback
        synchronized(p.lock) {
            for (i in 0 until count) {
                val next = (p.writeIndex + 1) % RING_FRAMES
                if (next == p.readIndex) break
                p.ring[p.writeIndex] = pullShorts.get(i)
                p.writeIndex = next
            }
        }
    }

    /** Stops the writer thread and releases the platform track. Safe to
     * call more than once; the drop safety net runs the same stop. */
    fun stop() = playback.stop()

    companion object {
        /** The lens mixer's fixed output format. */
        const val SAMPLE_RATE = 48_000
        private const val RING_FRAMES = 16_384
        private const val PULL_FRAMES = 4_096
        private const val CHUNK_FRAMES = 960
    }
}

File: sdk/swift/Sources/Gosslens/AudioOutput.swift
import AVFoundation

/// Routes the lens mixer to the device speaker. The engine's pull is
/// graph-thread only, so `pump()` runs in the frame loop and fills a
/// ring the audio thread's source node drains; an underrun plays
/// silence rather than blocking either side.
public final class GossAudioOutput {
    /// The lens mixer's fixed output format.
    public static let sampleRate: Double = 48_000
    private static let ringFrames = 16_384

    private let session: GossSession
    private let engine = AVAudioEngine()
    private var source: AVAudioSourceNode?
    private var ring = [Int16](repeating: 0, count: ringFrames)
    private var readIndex = 0
    private var writeIndex = 0
    private let lock = NSLock()
    private var pull = [Int16](repeating: 0, count: 4096)

    public init(session: GossSession) {
        self.session = session
    }

    /// Starts the platform audio engine with a mono source at the
    /// mixer rate; safe to call once after the session exists.
    public func start() throws {
        guard source == nil else { return }
        let format = AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: Self.sampleRate,
                                   channels: 1, interleaved: true)!
        let node = AVAudioSourceNode(format: format) { [weak self] _, _, frameCount, audioBufferList -> OSStatus in
            guard let self else { return noErr }
            let buffers = UnsafeMutableAudioBufferListPointer(audioBufferList)
            guard let raw = buffers[0].mData else { return noErr }
            let out = raw.bindMemory(to: Int16.self, capacity: Int(frameCount))
            self.lock.lock()
            for i in 0..<Int(frameCount) {
                if self.readIndex != self.writeIndex {
                    out[i] = self.ring[self.readIndex]
                    self.readIndex = (self.readIndex + 1) % Self.ringFrames
                } else {
                    out[i] = 0
                }
            }
            self.lock.unlock()
            return noErr
        }
        engine.attach(node)
        engine.connect(node, to: engine.mainMixerNode, format: format)
        do {
            try engine.start()
        } catch {
            engine.detach(node)
            throw error
        }
        source = node
    }

    /// Pulls the next mixer block into the ring. Call once per frame
    /// from the same thread that ticks the lens.
    public func pump(frames: UInt32 = 800) throws {
        let count = min(Int(frames), pull.count)
        try session.pullAudio(into: &pull, frames: UInt32(count))
        lock.lock()
        for i in 0..<count {
            let next = (writeIndex + 1) % Self.ringFrames
            if next == readIndex { break }
            ring[writeIndex] = pull[i]
            writeIndex = next
        }
        lock.unlock()
    }

    /// Stops the platform engine and detaches the source.
    public func stop() {
        if let node = source {
            engine.detach(node)
            source = nil
        }
        engine.stop()
    }

    deinit {
        stop()
    }
}
